src/app/clusters/thermostat-server/thermostat-server.cpp

```diff
@@ -699,10 +781,99 @@ CHIP_ERROR ThermostatAttrAccess::Write(const ConcreteDataAttributePath & aPath,
+        {
+            // Clear the pending schedules list
+            delegate->ClearPendingScheduleList();
+
+            Schedules::TypeInfo::DecodableType newSchedulesList;
+            ReturnErrorOnFailure(aDecoder.Decode(newSchedulesList));
+
+            // Iterate over the schedules and call the delegate to append to the list of pending schedules.
+            auto iter = newSchedulesList.begin();
+            while (iter.Next())
+            {
+                ScheduleStruct::DecodableType decodeSchedule = iter.GetValue();
+                ScheduleTransitionStruct::Type transitions[5];
+                auto iter2 = decodeSchedule.transitions.begin();
+                int i      = 0;
+
+                while (iter2.Next())
+                {
+                    transitions[i] = iter2.GetValue();
+                    i++;
+                }
+
+                ScheduleStruct::Type schedule;
+                schedule.systemMode     = decodeSchedule.systemMode;
+                schedule.scheduleHandle = decodeSchedule.scheduleHandle;
+                schedule.name           = decodeSchedule.name;
+                schedule.presetHandle   = decodeSchedule.presetHandle;
+                schedule.transitions    = DataModel::List<const Structs::ScheduleTransitionStruct::Type>(transitions, i);
+                schedule.builtIn        = decodeSchedule.builtIn;
+                ReturnErrorOnFailure(AppendPendingSchedule(delegate, schedule));
+            }
+            return iter.GetStatus();
+        }
+
+        // If the list operation is AppendItem, call the delegate to append the item to the list of pending schedules.
+        if (aPath.mListOp == ConcreteDataAttributePath::ListOperation::AppendItem)
+        {
+            ScheduleStruct::DecodableType decodeSchedule;
+            std::unique_ptr<ScheduleTransitionStruct::Type[]> transitions(
+                new ScheduleTransitionStruct::Type[delegate->GetNumberOfScheduleTransitions()]);
+
+            ReturnErrorOnFailure(aDecoder.Decode(decodeSchedule));
+
+            auto iter = decodeSchedule.transitions.begin();
+            int i     = 0;
+
+            while (iter.Next())
+            {
+                if (i >= delegate->GetNumberOfScheduleTransitions())
+                {
+                    ChipLogError(Zcl, "Too many transitions in schedule");
+                    return CHIP_IM_GLOBAL_STATUS(ConstraintError);
+                }
+
+                transitions[i] = iter.GetValue();
+                i++;
+            }
+
+            ScheduleStruct::Type schedule;
+            schedule.systemMode     = decodeSchedule.systemMode;
+            schedule.scheduleHandle = decodeSchedule.scheduleHandle;
+            schedule.name           = decodeSchedule.name;
+            schedule.presetHandle   = decodeSchedule.presetHandle;
+            schedule.transitions    = DataModel::List<const Structs::ScheduleTransitionStruct::Type>(transitions.get(), i);
```

Comment:
So why did we heap-allocate the transitions?  Couldn't we have just put them on the stack?